### DISCUSSION TITLE
Autenticathion issue

### DISCUSSION THREAD

**Discussion by valots12** [kubernetes-sigs/kwok#582] [Q&A] [answered]
Hi everyone,
I'm trying to getting familiar with the kwok tool. After installing it and switching the context to kwok-demo I receive a message error for every command, like getting nodes:

_error: You must be logged in to the server (Unauthorized)_

Can someone help me to solve this? 
Thanks in advance, I'm sorry but I'm very new in this env

**Comment by wzshiming** [comment]
Looks like a problem with the kubeconfig settings, Can you provide the steps to reproduce it?

**Comment by valots12** [comment]
Yeah sure. I followed the instruction for the installation of minikube and kwok and executed these commands:

_minikube start
kwokctl --name demo create cluster
kwokctl get clusters
kubectl config use-context kwok-demo
kubectl get nodes_

**Reply by wzshiming** [reply]
Why do you need to start minikube first?

https://kwok.sigs.k8s.io/docs/user/kwokctl-manage-cluster/
https://kwok.sigs.k8s.io/docs/user/kwok-manage-nodes-and-pods/

**Comment by valots12** [comment]
I would like to simulate a Argo workflow pipeline using fake nodes, don't I need to start minikube before?

**Reply by wzshiming** [reply]
I think you might need the kwok controller to use a fake node in a real node, which is different from kwokctl.

Deploy kwok in cluster https://kwok.sigs.k8s.io/docs/user/kwok-in-cluster/
Or running kwok out of cluster https://kwok.sigs.k8s.io/docs/user/kwok-out-cluster/

How to manage nodes and pods https://kwok.sigs.k8s.io/docs/user/kwok-manage-nodes-and-pods/

**Reply by valots12** [reply]
I actually don't plan to use real nodes. My idea is to create different fake environments for simulation with different memory and time limitations and extract the metrics in different conditions. I already have a workflow in Argo (works fine in k8s), now I need to submit it in kwok to see how it works.

**Reply by wzshiming** [reply]
Cool, that just use kwokctl to create the cluster

https://kwok.sigs.k8s.io/docs/user/kwokctl-manage-cluster/
https://kwok.sigs.k8s.io/docs/user/kwok-manage-nodes-and-pods/

![](https://kwok.sigs.k8s.io/img/demo/manage-clusters.svg)

**Reply by valots12** [reply]
Thanks, I already looked at the documentation. The problem is that without starting minikube I cannot create a node since it is not connected and I get a connection to the server error:

_Unable to connect to the server: no route to host_

**Reply by wzshiming** [reply]
Please get a new machine and download the latest version and try again, I suspect it's a problem with your environment.

**Reply by valots12** [reply]
Ok, I'm gonna try it.

**Reply by valots12** [reply]
Thanks for the support, it seems like it is working properly now. I can manage correctly nodes and pods, how can I submit a workflow in a specific cluster that I have?

**Reply by wzshiming** [reply]
> how can I submit a workflow in a specific cluster that I have?

I don't understand what you mean, can you make it clear?

**Reply by valots12** [reply]
I mean, usually I use this command for executing a workflow in k8s:

_argo submit -n argo --watch {yaml_path}_

How can I run the same within the fake environment to simulate it? And also, can I extract metrics like CPU or memory consumption from it?

**Reply by wzshiming** [reply]
> How can I run the same within the fake environment to simulate it? And also, can I extract metrics like CPU or memory consumption from it?

kwokctl doesn't actually run any containers, anything that requires a real pod needs to rely on Kind

1. Create kind cluster with kwok

``` bash
kwokctl create cluster --runtime kind
```

2. Deploy argo controller in kind's node
3. argo submit -n argo --watch {yaml_path}

**Reply by valots12** [reply]
what to you mean for the second step?

**Reply by wzshiming** [reply]
> what to you mean for the second step?

Install argocd in kind's node

https://argo-cd.readthedocs.io/en/stable/getting_started/#1-install-argo-cd

**Reply by valots12** [reply]
I think that Argo CD does not support the submit command, maybe you are referring to Argo Workflows?

**Reply by wzshiming** [reply]
> maybe you are referring to Argo Workflows?

Yeah

**Reply by valots12** [reply]
ok good. What about the fake nodes and pods? Do I have to create them inside whis new Kind cluster?

**Reply by wzshiming** [reply]
yes

**Reply by valots12** [reply]
can I ask you why do I need a kind cluster and not a kwok one? Thanks for everything btw

**Reply by wzshiming** [reply]
The main job of Kwok is a behavioral simulation, which is mainly Node and Pod. so a controller deployed on a kwok node won't really work and won't be able to handle CRD. this requires a way to run the controller, and Kind is a great addition.

**Reply by valots12** [reply]
and it's not possible to simulate resources just using Kind, is it correct?

**Reply by wzshiming** [reply]
You can install a Kwok controller in a Kind cluster or run a Kwok controller outside the cluster to simulate resources.

https://kwok.sigs.k8s.io/docs/user/kwok-in-cluster/
https://kwok.sigs.k8s.io/docs/user/kwok-out-cluster/

**Reply by valots12** [reply]
ok, thanks a lot

**Reply by valots12** [reply]
I made some progress, but I still have some doubts about kwok. After creating a kind cluster and installing argo workflows, it's easy to submit a pipeline using the argo controller. In order to use fake resources, is it correct if I install kwok in the cluster and create a fake node in the argo namespace? Then the missing step would be the integration between argo and kwok controllers.

**Comment by wzshiming** [comment, marked as answer]
Follow the steps below and I guess it should be what you want

## Set up cluster

``` bash
wget https://github.com/kubernetes-sigs/kwok/raw/main/stages/pod-fast.yaml
sed 's/Job/Workflow/g' pod-fast.yaml > workflow-fast.yaml
kwokctl create cluster --runtime kind -c workflow-fast.yaml
```

## Deploy argo

``` bash
kubectl create namespace argo
kubectl apply -n argo -f https://github.com/argoproj/argo-workflows/releases/download/v3.4.8/install.yaml
kubectl patch deploy argo-server -n argo --type=json -p='[{"op":"add","path":"/spec/template/spec/nodeName","value":"kwok-kwok-control-plane"}]'
kubectl patch deploy workflow-controller -n argo --type=json -p='[{"op":"add","path":"/spec/template/spec/nodeName","value":"kwok-kwok-control-plane"}]'
```

## Create node

``` bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: Node
metadata:
  annotations:
    node.alpha.kubernetes.io/ttl: "0"
    kwok.x-k8s.io/node: fake
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/os: linux
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: kwok-node-0
    kubernetes.io/os: linux
    kubernetes.io/role: agent
    node-role.kubernetes.io/agent: ""
    type: kwok
  name: kwok-node-0
status:
  allocatable:
    cpu: 32
    memory: 256Gi
    pods: 110
  capacity:
    cpu: 32
    memory: 256Gi
    pods: 110
EOF
```

## Test

``` bash
argo submit -n argo --watch https://raw.githubusercontent.com/argoproj/argo-workflows/master/examples/hello-world.yaml
```

``` log
Name:                hello-world-p9g6d
Namespace:           argo
ServiceAccount:      unset (will run with the default ServiceAccount)
Status:              Succeeded
Conditions:          
 PodRunning          False
 Completed           True
Created:             Wed Jun 14 12:02:21 +0800 (10 seconds ago)
Started:             Wed Jun 14 12:02:21 +0800 (10 seconds ago)
Finished:            Wed Jun 14 12:02:31 +0800 (now)
Duration:            10 seconds
Progress:            1/1
ResourcesDuration:   0s*(100Mi memory),0s*(1 cpu)

STEP                  TEMPLATE  PODNAME            DURATION  MESSAGE
 ✔ hello-world-p9g6d  whalesay  hello-world-p9g6d  1s       
```

``` log
kubectl get pod -n argo -o wide
NAME                                   READY   STATUS      RESTARTS      AGE     IP           NODE                      NOMINATED NODE   READINESS GATES
argo-server-74988d78f-5mvls            1/1     Running     0             16m     10.244.0.6   kwok-kwok-control-plane   <none>           <none>
hello-world-lsznk                      0/2     Completed   0             10m     10.244.1.0   kwok-node-0               <none>           <none>
workflow-controller-786dc8cdb6-7g56s   1/1     Running     2 (12m ago)   16m     10.244.0.5   kwok-kwok-control-plane   <none>           <none>
```

**Reply by valots12** [reply]
Thanks, it works. I'll look more in details at the code to understand it better.

**Reply by valots12** [reply]
Hi, it seems that is working properly only with pipelines that do not require artifacts between steps. Is it correct?

**Reply by wzshiming** [reply]
I haven't actually used argo, can you provide a minimal example of how to reproduce the problem?

**Reply by valots12** [reply]
Sure. If you submit with argo this yaml file you should be able to reproduce my error: https://github.com/argoproj/argo-workflows/blob/master/examples/artifact-passing.yaml

**Reply by wzshiming** [reply]
add optional for all artefacts, will work

``` patch
@@ -16,6 +16,7 @@
             arguments:
               artifacts:
                 - name: message
+                  optional: true
                   from: "{{steps.generate-artifact.outputs.artifacts.hello-art}}"
 
     - name: whalesay
@@ -27,12 +28,14 @@
         artifacts:
           - name: hello-art
             path: /tmp/hello_world.txt
+            optional: true
 
     - name: print-message
       inputs:
         artifacts:
           - name: message
             path: /tmp/message
+            optional: true
       container:
         image: alpine:latest
         command: [sh, -c]
```

**Reply by valots12** [reply]
thanks, it is working now. can I ask you how the artifacts are handled with this optional command?

**Reply by wzshiming** [reply]
There are no artifacts, all are fake. 😄

**Reply by valots12** [reply]
Yes of course 😄 but how can the workflow be simulated then?

**Reply by valots12** [reply]
is it the same of running a yaml without artifacts?

**Reply by wzshiming** [reply]
Hmmm, I'm not sure about the details of argo, maybe you need to ask in their community.
